File: api/src/schema.ts
export const WINE_TYPES = ["red", "white", "rosé", "orange", "sparkling", "dessert"] as const;
export type WineType = (typeof WINE_TYPES)[number];

export const SOURCE_MATERIALS = ["kit", "juice_bucket", "fresh_grapes"] as const;
export type SourceMaterial = (typeof SOURCE_MATERIALS)[number];

export const BATCH_STAGES = [
  "must_prep",
  "primary_fermentation",
  "secondary_fermentation",
  "stabilization",
  "bottling",
] as const;
export type BatchStage = (typeof BATCH_STAGES)[number];

export const ALL_STAGES = [
  "receiving", "crushing", "must_prep",
  "primary_fermentation", "pressing",
  "secondary_fermentation", "malolactic",
  "stabilization", "fining", "bulk_aging", "cold_stabilization", "filtering",
  "bottling", "bottle_aging",
] as const;
export type AllStage = (typeof ALL_STAGES)[number];

export const BATCH_STATUSES = ["active", "completed", "archived", "abandoned"] as const;
export type BatchStatus = (typeof BATCH_STATUSES)[number];

export const ACTIVITY_TYPES = ["addition", "racking", "measurement", "tasting", "note", "adjustment"] as const;
export type ActivityType = (typeof ACTIVITY_TYPES)[number];

export const OAK_TYPES = ["none", "american", "french", "hungarian"] as const;
export type OakType = (typeof OAK_TYPES)[number];

export const OAK_FORMATS = ["barrel", "chips", "cubes", "staves", "spiral"] as const;
export type OakFormat = (typeof OAK_FORMATS)[number];

export const MLF_STATUSES = ["not_planned", "pending", "in_progress", "complete"] as const;
export type MlfStatus = (typeof MLF_STATUSES)[number];

export const WAYPOINT_ALLOWED_STAGES: Record<BatchStage, readonly AllStage[]> = {
  must_prep: ["receiving", "crushing", "must_prep"],
  primary_fermentation: ["primary_fermentation", "pressing"],
  secondary_fermentation: ["secondary_fermentation", "malolactic"],
  stabilization: ["stabilization", "fining", "bulk_aging", "cold_stabilization", "filtering"],
  bottling: ["bottling", "bottle_aging"],
};

export const SO2_CHEMICALS = ["K2S2O5", "SO2", "Campden", "K-meta", "Potassium metabisulfite"] as const;
export const MLF_CULTURES = ["MLB", "Leuconostoc", "CH16", "VP41", "malolactic"] as const;

export const WAYPOINT_ORDER: readonly BatchStage[] = BATCH_STAGES;

File: api/src/schema.py
"""D1/SQLite schema reference. The SQL migration files are the source of truth.
This module provides the schema SQL for use in tests."""

from pathlib import Path

MIGRATIONS_DIR = Path(__file__).parent.parent / "migrations"

BATCH_STAGES = (
    "must_prep",
    "primary_fermentation",
    "secondary_fermentation",
    "stabilization",
    "bottling",
)

ALL_STAGES = (
    "receiving", "crushing", "must_prep",
    "primary_fermentation", "pressing",
    "secondary_fermentation", "malolactic",
    "stabilization", "fining", "bulk_aging", "cold_stabilization", "filtering",
    "bottling", "bottle_aging",
)

WINE_TYPES = ("red", "white", "rosé", "orange", "sparkling", "dessert")

SOURCE_MATERIALS = ("kit", "juice_bucket", "fresh_grapes")

BATCH_STATUSES = ("active", "completed", "archived", "abandoned")

ACTIVITY_TYPES = ("addition", "racking", "measurement", "tasting", "note", "adjustment")

# Map batch waypoints to allowed activity stages
WAYPOINT_ALLOWED_STAGES: dict[str, tuple[str, ...]] = {
    "must_prep": ("receiving", "crushing", "must_prep"),
    "primary_fermentation": ("primary_fermentation", "pressing"),
    "secondary_fermentation": ("secondary_fermentation", "malolactic"),
    "stabilization": ("stabilization", "fining", "bulk_aging", "cold_stabilization", "filtering"),
    "bottling": ("bottling", "bottle_aging"),
}

# Ordered waypoints for /advance
WAYPOINT_ORDER = list(BATCH_STAGES)


def get_migration_sql() -> str:
    """Read the initial migration SQL for use in tests."""
    migration_file = MIGRATIONS_DIR / "0001_initial.sql"
    return migration_file.read_text()
